### PR TITLE
PWA Support for Angular in CLI

### LINKED ISSUE #12180: Optional PWA support on new solutions for Angular UI
Part of https://github.com/abpframework/abp/issues/11741

### PR DESCRIPTION
**Pull request by enisn**
Closes https://github.com/abpframework/abp/issues/12180

### CODE REVIEW

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        ICmdHelper cmdHelper,
+        PackageJsonFileFinder packageJsonFileFinder)
+    {
+        CmdHelper = cmdHelper;
+        PackageJsonFileFinder = packageJsonFileFinder;
+    }
+
+    public virtual void AddPwaSupport(string rootDirectory)
+    {
+        var fileList = PackageJsonFileFinder.Find(rootDirectory).Where(x => x.Contains("angular")).ToList();
```
Comment (yekalkan) [inline framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: It's better to check if it is next to a `angular.json` file.

PackageJsonFileFinder.Find(rootDirectory).Where(x => File.Exist( x.RemovePostFix("package.json") + "angular.json") ).ToList();

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        if (!fileList.Any())
```
Comment (yekalkan) [inline framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: I don't think this control is necessary.

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        fileList.ForEach(AddPwaSupportToProject);
+    }
+
+    protected virtual void AddPwaSupportToProject(string filePath)
+    {
+        var directory = Path.GetDirectoryName(filePath).EnsureEndsWith(Path.DirectorySeparatorChar);
+
+        CmdHelper.RunCmd($"cd {directory} && ng add @angular/pwa --skip-confirmation");
```
Comment (yekalkan) [inline framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: It is better to use `workingDirectory` parameter:

CmdHelper.RunCmd("ng add @angular/pwa --skip-confirmation", workingDirectory: directory);

**Review by yekalkan — CHANGES_REQUESTED** [review]

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs (hunk 1 above)
Comment (enisn) [reply framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: it's folder, So I added to `Path.DirectorySeparatorChar` to the end of it

Comment (yekalkan) [reply framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: No, you misunderstood me. You need to check if there is a `angular.json` file in the same folder, to understand if it is an angular project or not.

image

Comment (enisn) [reply framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: Ok got it 👍

Comment (enisn) [reply framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: You can check commit https://github.com/abpframework/abp/commit/d30082bc9616c2f61aa7319aa9646a86401b79ad

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        ICmdHelper cmdHelper,
+        PackageJsonFileFinder packageJsonFileFinder)
+    {
+        CmdHelper = cmdHelper;
+        PackageJsonFileFinder = packageJsonFileFinder;
+    }
+
+    public virtual void AddPwaSupport(string rootDirectory)
+    {
+        var fileList = PackageJsonFileFinder.Find(rootDirectory).Where(x => x.Contains("angular.json")).ToList();
```
Comment (yekalkan) [inline framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: This won't work. `x` is the full path of `appsettings.json` file, you shouldn't filter it with `angular.json`.

```suggestion
       var fileList = PackageJsonFileFinder.Find(rootDirectory).Where(x => File.Exist( x.RemovePostFix("package.json") + "angular.json") ).ToList();
```

**Review by yekalkan — CHANGES_REQUESTED** [review]

File: framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs (hunk 1 above)
Comment (enisn) [reply framework/src/Volo.Abp.Cli.Core/Volo/Abp/Cli/ProjectModification/AngularPwaSupportAdder.cs]: I see,

I've tested now, this way works well 👍 
https://github.com/abpframework/abp/commit/e6e5cb05a8aeb6222933f0977ef25e4cf8070399

**Review by yekalkan — APPROVED** [review]

**Pull request merged**